pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -220,6 +258,13 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
     Comparable newComparisonValue = recordInfo.getComparisonValue();
     _primaryKeyToRecordLocationMap.compute(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction),
         (primaryKey, currentRecordLocation) -> {
+          // Update the largestSeenComparisonValueMs when add new record. If records during addSegments has a newer
+          // comparison column values than addRecords, it's a bug and should not happen.
+          if (_upsertTTLInComparisonTimeUnit > 0) {
+            if (recordInfo.getComparisonValue().compareTo(_largestSeenComparisonValue) > 0) {
+              _largestSeenComparisonValue = (double) recordInfo.getComparisonValue();
```

Comment:
The cast can throw exception. Please add a test with non-double comparison column